extensions/coap/controllerservice/CoapConnector.cpp

```diff
@@ -0,0 +1,189 @@
+
+void CoapConnectorService::initialize() {
+  if (initialized_)
+    return;
+
+  callback_pointers ptrs;
+  ptrs.data_received = receiveMessage;
+  ptrs.received_error = receiveError;
+  init_coap_api(this, &ptrs);
+
+  std::lock_guard<std::mutex> lock(initialization_mutex_);
+
+  ControllerService::initialize();
+
+  initializeProperties();
+
+  initialized_ = true;
+}
+
+void CoapConnectorService::onEnable() {
+  std::string port_str;
+  if (getProperty(RemoteServer.getName(), host_) && !host_.empty() && getProperty(Port.getName(), port_str) && !port_str.empty()) {
+    core::Property::StringToInt(port_str, port_);
+  } else {
+    // this is the case where we aren't being used in the context of a single controller service.
+    if (configuration_->get("nifi.c2.agent.coap.host", host_) && configuration_->get("nifi.c2.agent.coap.port", port_str)) {
+      core::Property::StringToInt(port_str, port_);
+    }
+
+  }
+}
+
+CoapConnectorService::CoAPMessage CoapConnectorService::sendPayload(uint8_t type, const std::string endpoint, unsigned char *payload, size_t size) {
+  struct coap_context_t* ctx=NULL;
+  struct coap_session_t* session=NULL;
+
+  coap_address_t dst_addr, src_addr;
+  coap_uri_t uri;
+  uri.host.s = reinterpret_cast<unsigned char *>(const_cast<char*>(host_.c_str()));
+  uri.host.length = host_.size();
+  uri.path.s = reinterpret_cast<unsigned char *>(const_cast<char*>(endpoint.c_str()));
+  uri.path.length = endpoint.size();
+  uri.port = port_;
+
+  fd_set readfds;
+  coap_pdu_t* request;
+  unsigned char get_method = 1;
+
+  int res = resolve_address(&uri.host, &dst_addr.addr.sa);
+  if (res < 0) {
+    return CoAPMessage();
+  }
+
+  dst_addr.size = res;
+  dst_addr.addr.sin.sin_port = htons(uri.port);
+
+  void *addrptr = NULL;
+  char port_str[NI_MAXSERV] = "0";
+  char node_str[NI_MAXHOST] = "";
+  switch (dst_addr.addr.sa.sa_family) {
+    case AF_INET:
+      addrptr = &dst_addr.addr.sin.sin_addr;
+
+      /* create context for IPv4 */
+      if (!create_session(&ctx, &session, node_str[0] == 0 ? "0.0.0.0" : node_str, port_str, &dst_addr)) {
+        break;
+      } else {
+        return CoAPMessage();
+      }
+    case AF_INET6:
+      addrptr = &dst_addr.addr.sin6.sin6_addr;
+
+      /* create context for IPv6 */
+      if (!create_session(&ctx, &session, node_str[0] == 0 ? "::" : node_str, port_str, &dst_addr)) {
+        break;
+      } else {
+        return CoAPMessage();
+      }
+    default:
+      ;
+  }
+
+  // we want to register handlers in the event that an error occurs or nack is returned
+  // from the library
+  coap_register_event_handler(ctx, coap_event);
+  coap_register_nack_handler(ctx, no_acknowledgement);
+
+  coap_context_set_keepalive(ctx, 1);
+
+  coap_str_const_t pld;
+  pld.length = size;
+  pld.s = payload;
```

Comment:
Given .s is const uint8_t* ( https://github.com/obgm/libcoap/blob/develop/include/coap2/str.h ), could you change the signature of the function to handle the payload as const?

Reply:
I'll have to look what it is now, but we'll probably keep with that the outer facing types are and translate as needed, but, and I don't know if we can move to const as we'll probably be taking ownership of that buffer in the next commit.